### PR TITLE
Add TW3Cast results

### PR DESCRIPTION
**Pull request by nattanioCalor**
Adds results for TW3Cast (TW3 Partners): a frozen per-configuration router over lightly fine-tuned experts (Chronos-2, TiRex, Toto 2.5B LoRA), no agent and no LLM at inference, no test-data leakage in any component. Scores computed with the official harness (gluonts evaluate_forecasts, leaderboard parameters). Code and router table: https://github.com/TW3-Partners-OS/TW3-Cast

### CODE REVIEW

**Comment by cuthalionn** [comment]
Hi @nattanioCalor, 

Thanks for the submission! Frankly this is the first time we get a mixture of routing + finetuning submission so I am not sure about what to label it, but `agentic` seems like a better fit to me since it is in the end routing multiple models and seems like a fairer comparison to other models in the same family in the leaderboard. Would you be okay with making that change? 

FYI: I started working on replicating the results.

**Comment by cuthalionn** [comment]
Hi @nattanioCalor, 

Separately on replication I tried to give it a go on a smaller subset and realised that certain parts of the code is not released? For bizitobs_application/10S/{medium,long}, the public router specifies __tournament__,  but the repo does not provide the  tournament code and winners are not published (please correct me if I am wrong). Any suggestions how to proceed?

Moreover I tried m4 weekly which  does not require a tournament:

| Model | Dataset/frequency/term | Rep MASE | Submitted MASE | Δ MASE | Rep WSQL | Submitted WSQL | Δ WSQL |
|---|---|---:|---:|---:|---:|---:|---:|
| TW3Cast | m4_weekly/W/short | 1.906935 | 1.869772 | +0.037163 | 0.035549 | 0.035317 | +0.000232 |


I followed the public TW3-Cast README and used the published expert checkpoints and router configuration, I used the published frozen Toto/TimesFM route and got ~0.03 diff.

Would appreciate your help to address these problems. It would make life much easier for me if the repo had all code available and a single entrypoint that lets me replicate any dataset.  Thanks again!

**Comment by nattanioCalor** [comment]
Hi @cuthalionn, thanks for the careful review and for taking the time to replicate.

**On the label:** no objection, `agentic` is a reasonable fit since the system does route over multiple models, and it puts us in the right comparison group. `config.json` is updated to `"model_type": "agentic"`. For the record, TW3Cast runs no agent and no language model, and makes no decision at inference time: the routing table is computed once on the training split and then frozen.

**On the tournament and replication:** you were right that this was not verifiable. The full replication pack is now released:

- `served/<config>.npy` — the served forecast quantiles for all 97 configurations, tournament-mode ones included, shape `(n_windows, 9, horizon)`: the exact arrays behind the submission.
- `decisions/tournament_decisions.csv` — for each tournament-served configuration, the member selected on the train-side backtest and the number of per-window gate switches.
- A single entrypoint, as you asked:

```bash
python replicate.py --config "m4_weekly/W/short"   # or --all
```

It downloads the served quantiles from the model release, evaluates them with the official harness and checks every metric against the submitted file. On `m4_weekly/W/short` it reports `EXACT MATCH` on all 11 metrics, and likewise on tournament-served configurations such as `bizitobs_application/10S/medium`.

**On the ~0.03 MASE difference you measured:** that gap is real and expected from `predict.py`. It recomputes forecasts from the released checkpoints using each architecture's reference recipe, but the submitted lines were served with configuration-specific settings (8192-token context, the official Toto wrapper on some configurations, joint multivariate prediction for multi-channel datasets). Recomputation therefore lands within about 0.3 % of the served line rather than on it. Exact replication goes through the served arrays; `predict.py` remains the reference for recomputing forecasts from weights.

**One more thing, for transparency:** while preparing this, I found that 17 of our 97 lines came from an earlier build whose source artifacts had been lost, and could not be reproduced. I re-served them from the artifacts named in our manifest and updated the submission with what I can actually measure. This costs a little accuracy — mean MASE rank moves from 19.41 to 19.56 — but every line in the updated file is now reproducible from the released material.

Repository: https://github.com/TW3-Partners-OS/TW3-Cast
Replication pack: https://huggingface.co/TW3PartnersLLM/TW3Cast

Happy to add anything else that would help the review.

**Comment by nattanioCalor** [comment]
Hi @cuthalionn, thanks for the careful review and for taking the time to replicate.

**On the label:** no objection, `agentic` is a reasonable fit since the system does route over multiple models, and it puts us in the right comparison group. `config.json` is updated to `"model_type": "agentic"`. For the record, TW3Cast runs no agent and no language model, and makes no decision at inference time: the routing table is computed once on the training split and then frozen.

**On the tournament:** you are right, and here is where things stand. The *method* is documented in [REPORT.md](https://github.com/TW3-Partners-OS/TW3-Cast/blob/main/REPORT.md) (section 4): pool members are scored on a backtest carved from the training split, the winner is selected under a dual MASE-plus-CRPS criterion with an asymmetric margin against locally trained candidates, near-ties fall back to the heavyweight blend, and a conservative per-window gate may reroute individual windows at prediction time. What was missing is everything needed to check it, so both the decisions and the outputs are now published; the selection code itself is not released, but nothing in the verification depends on it.

**On replication:** the full pack is now released:

- `served/<config>.npy` — the served forecast quantiles for all 97 configurations, tournament-mode ones included, shape `(n_windows, 9, horizon)`: the exact arrays behind the submission.
- `decisions/tournament_decisions.csv` — for each tournament-served configuration, the member selected on the train-side backtest and the number of per-window gate switches.
- A single entrypoint, as you asked:

```bash
python replicate.py --config "m4_weekly/W/short"   # or --all
```

It downloads the served quantiles from the model release, evaluates them with the official harness and checks every metric against the submitted file. On `m4_weekly/W/short` it reports `EXACT MATCH` on all 11 metrics, and likewise on tournament-served configurations such as `bizitobs_application/10S/medium`.

**On the ~0.03 MASE difference you measured:** that gap is real and expected from `predict.py`. It recomputes forecasts from the released checkpoints using each architecture's reference recipe, but the submitted lines were served with configuration-specific settings (8192-token context, the official Toto wrapper on some configurations, joint multivariate prediction for multi-channel datasets). Recomputation therefore lands within about 0.3 % of the served line rather than on it. Exact replication goes through the served arrays; `predict.py` remains the reference for recomputing forecasts from weights.

**One more thing, for transparency:** while preparing this, I found that 17 of our 97 lines came from an earlier build whose source artifacts had been lost, and could not be reproduced. I re-served them from the artifacts named in our manifest and updated the submission with what I can actually measure. This costs a little accuracy — mean MASE rank moves from 19.41 to 19.56 — but every line in the updated file is now reproducible from the released material.

Repository: https://github.com/TW3-Partners-OS/TW3-Cast
Replication pack: https://huggingface.co/TW3PartnersLLM/TW3Cast

Happy to add anything else that would help the review.

**Comment by cuthalionn** [comment]
Hi @nattanioCalor, thanks for the update I am able to replicate results with the new changes! The leaderboard is updated with these results.
